packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1362,18 +1340,22 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         return context;
     }
 
-    public async _createDataStoreWithProps(pkg: string | string[], props?: any, id = uuid()):
-        Promise<IFluidDataStoreChannel> {
-        return this._createFluidDataStoreContext(
-            Array.isArray(pkg) ? pkg : [pkg], id, false /* isRoot */, props).realize();
+    public async _createDataStoreWithProps(pkg: string | string[], props?: any, id = uuid()): Promise<IFluidRouter> {
+        return this._createDataStore(Array.isArray(pkg) ? pkg : [pkg], false /* isRoot */, id, props);
     }
 
     private async _createDataStore(
         pkg: string | string[],
         isRoot: boolean,
         id = uuid(),
-    ): Promise<IFluidDataStoreChannel> {
-        return this._createFluidDataStoreContext(Array.isArray(pkg) ? pkg : [pkg], id, isRoot).realize();
+        props?: any,
+    ): Promise<IFluidRouter> {
+        const context = this._createFluidDataStoreContext(Array.isArray(pkg) ? pkg : [pkg], id, isRoot, props);
+        const fluidDataStore = await context.realize();
```

Comment:
>const fluidDataStore = await context.realize(); [](start = 8, length = 47)

should this realization move in the if as we only need to realize if it's a root

Reply:
I'd say that all creation paths should create objects right away, as that's the right place to pay the cost and surface creation errors. At least that's the stance we are taking in Aqueduct level (as doing it otherwise was causing buts, including DDSs not being created in attach snapshots). Here we can do either way, but it feels right to pay the cost upfront (otherwise caller should not create data store :)
That's not primary reason, but requests, unfortunately, do not use JS exceptions as a way to communicate errors, and it's painful to diagnose them.
That all said, I'm not set on it - either way will work fine, I believe